Collapse help into expandable command topics

u2mejc opened the pull request:
This resolves #1688 :+1: 

```
$ dokku
Usage: dokku [--quiet|--trace|--rm-container|--rm|--force] COMMAND <app> [command-specific-options]

Primary help options, type "dokku COMMAND:help" for more details, or dokku help --all to see all commands.
Commands:
    apps             List your apps
    checks           Show zero-downtime status
    config           Display all global or app-specific config vars
    docker-options   Display app's Docker options for all phases (or comma separated phase list)
    domains          List domains
    enter            Connect to a specific app container
    events           Show the last events (-t follows)
    logs             Show the last logs for an application
    ls               Pretty listing of deployed applications and containers
    plugin           Print active plugins
    proxy            Show proxy for app
    run              Run a command in the environment of an application
    tags             List all app image tags
    trace            Enable dokku tracing
    url              Show the first URL for an application (compatibility)
    urls             Show all URLs for an application
    version          Print dokku's version

$ dokku domains:help
    domains:add <app> DOMAIN      Add a domain to app
    domains [<app>]               List domains
    domains:clear <app>           Clear all domains for app
    domains:disable <app>         Disable VHOST support
    domains:enable <app>          Enable VHOST support
    domains:remove <app> DOMAIN   Remove a domain from app
    domains:set-global <domain>   Set global domain name
```

michaelshobbs commented:
Can you show what the outputs look like in the description of the PR?

michaelshobbs commented on dokku:
```diff
@@ -159,11 +159,16 @@ case "$1" in
     echo ""
-    echo "Options:"
+    echo "Primary help options, type \"dokku COMMAND:help\" for more details, or dokku help --all to see all commands."
+    echo "Commands:"
```
I think consolidating lines 162 and 163 to the following (which is an almost direct copy of heroku's msg) makes things clearer.
```
    echo "Primary help topics, type 'dokku TOPIC:help' for more details:"
```

u2mejc replied:
You've obscured the --all, I added that feature for you. Please advise. :wink:

michaelshobbs commented:
@u2mejc Also, I have thoughts on how to do this if you'd like to discuss offline.

michaelshobbs commented on plugins/docker-options/commands:
```diff
@@ -4,12 +4,15 @@ set -eo pipefail; [[ $DOKKU_TRACE ]] && set -x
-    docker-options <app> <phase(s)>, Display app's Docker options for phase (comma separated phase list)
+    help_content='    docker-options <app>, Display app'"'"'s Docker options for all phases
+    docker-options <app> <phase(s)>, Display app'"'"'s Docker options for phase (comma separated phase list)
     docker-options:add <app> <phase(s)> OPTION, Add Docker option to app for phase (comma separated phase list)
```
That's not a helpful response. Can you clarify please?

u2mejc replied:
ah sorry about that, it's because that's a separate issue about POSIX argument syntax standards compliance. We have something of a history of taking things into the weeds on minor points like that and I was hoping to not touch the topic (since that was preexisting before this PR).

michaelshobbs replied:
I think it's fine to refactor this bit in the context of cleaner help output.

u2mejc commented:
> Any way to get the bare method (domains in this case) above other subcommands? Just looks stupid atm.

Was discussed here: https://github.com/dokku/dokku/pull/1830 
> michaelshobbs commented 8 days ago
> It's not that simple since some plugin commands do something functional when no args are given.

michaelshobbs commented:
I don't think that's the same context. I believe @josegonzalez is asking about sort order of the sub topic output. i.e.

note that the bare `dokku domains` command is first.
```
$ dokku domains:help
    domains [<app>]               List domains
    domains:add <app> DOMAIN      Add a domain to app
    domains:clear <app>           Clear all domains for app
    domains:disable <app>         Disable VHOST support
    domains:enable <app>          Enable VHOST support
    domains:remove <app> DOMAIN   Remove a domain from app
    domains:set-global <domain>   Set global domain name
```

josegonzalez commented:
@michaelshobbs that "additional commands" section on the bare commands could break parsing output.

michaelshobbs commented:
Parsing how? Does help output need to be machine readable?

josegonzalez commented:
No but we can't expect to parse:

```
dokku apps
 === My Apps
 example
 example2

 === Collaborated Apps
 theirapp   other@owner.name

Additional commands, type "heroku help COMMAND" for more details:

  apps:create [NAME]             #  create a new app
  apps:destroy --app APP         #  permanently destroy an app
  apps:info                      #  show detailed app information
  apps:join --app APP            #  add yourself to an organization app
  apps:leave --app APP           #  remove yourself from an organization app
  apps:lock --app APP            #  lock an organization app to restrict access
  apps:open --app APP            #  open the app in a web browser
  apps:rename NEWNAME --app APP  #  rename the app
  apps:transfer RECIPIENT        #  transfer an app to another user or organization
  apps:unlock --app APP          #  unlock an organization app so that any org member can join it
```

josegonzalez commented:
I just don't want the extra help output on things like `dokku apps`. I'm agreeing with everything else.

michaelshobbs commented:
Oh I'm not saying the `dokku apps` command should output help. Instead `dokku apps:help` would output this screen.

josegonzalez commented:
Ah okay, thats fine by me :)

michaelshobbs commented:
**Top level help output**
```
Usage: dokku [--quiet|--trace|--rm-container|--rm|--force|--help] COMMAND <app> [command-specific-options]

Primary help topics, type "dokku TOPIC:help" for more details:
    apps                # manage apps (create, destroy)
    checks              # manage zero downtime features
    config              # manage global or app-specific configuration
    docker-options      # manage docker options for all phases
    domains             # manage domains
    enter               # connect to a specific app container
    logs                # display logs for an app
    proxy               # manage proxy configuration
    run                 # run a command in the environment of an app
    tags                # manage app image tags

Additional topics:
    events              # display events
    ls                  # display deployed apps and containers
    plugin              # manage plugins
    trace               # enable debug tracing
    version             # display version
    urls                # display all URLs for an app

3rd party plugin topics:
  deploymentkeys:create <app>   Create a pair of app-specific deployment keys
  deploymentkeys:delete <app>   Delete the current pair of deployment keys
  deploymentkeys:shared         Shows the current shared public key
  deploymentkeys:show   <app>   Shows the current public key to add to your VCS
```

**Core topic detail**
```
Usage: dokku apps

 list your apps

Example:

 $ dokku apps
=====> My Apps
 example
 example2

Additional commands:

  apps                              # list your apps
  apps:create <app>                 # create a new app
  apps:destroy <app>                # permanently destroy an app
  apps:rename <old-app> <new-app>   # rename an app
```

Notes
- some of the top level copy has changed
- there is a new 3rd party section for non-core help output
- we'll want to come up with a clean API to give to 3rd party devs for help output (something like define a usage() function with a given text format and we'll handle the collation)
 
I'm sure I'm missing something here so feedback is welcome

josegonzalez commented:
I don't want to do this:

```
 $ dokku apps
=====> My Apps
 example
 example2

Additional commands:

  apps                              # list your apps
  apps:create <app>                 # create a new app
  apps:destroy <app>                # permanently destroy an app
  apps:rename <old-app> <new-app>   # rename an app
```

For bare commands that already have output, adding more things to the end seems... wrong?

michaelshobbs commented:
Ok were confused again. I'm not saying to change the functionality of `dokku apps`. This is only the output of `dokku apps:help`

josegonzalez commented:
Yeah but the command you have up there is `dokku apps` and you're appending `dokku apps:help` to it. I don't want to append the help output to commands that already do something.

michaelshobbs commented:
I'm not following. 

`dokku apps` does what it always did. Help will not be shown in any other case than `dokku TOPIC:help`

josegonzalez commented:
Ah shit, I didn't realize that whole block was the output of `dokku app:help`. Fuck, my bad.

u2mejc commented:
How do you guys feel about me drawing a hardline between "core" and other plugins in the file structure? In the Makefile (haven't looked at the deb) we symlink core plugins to both directories which is breaking my devision here. Cool if we respect the divide as: if it's in this repo, it's core, if not, then it's not core?

michaelshobbs commented:
My thinking was that core plugins are those in `/var/lib/dokku/core-plugins`. I don't think that includes official plugins like MySQL, PostgreSQL, etc. Is that true? 😁

josegonzalez commented:
Fine by me. @michaelshobbs did similar with the `--app` flag in the commands refactor.

josegonzalez commented:
Correct, those are official, but not core.

u2mejc commented:
cool, we're on the same page.

u2mejc commented:
I split these into two commits, and I think it shouldn't be squashed, but we can move to another PR if needed.

This follows up the work of dividing the "community" (don't say third party) plugin paths. Specifically, when the Makefile was updated to no long insert symlinks in /var/lib/dokku/plugin, it removed the assumption that all plugins could be in either dir. After following some other avenues, I decided to simply wrap plugn in a function (since it relies on a global var).

michaelshobbs commented on plugins/ps/commands:
```diff
@@ -4,8 +4,7 @@ set -eo pipefail; [[ $DOKKU_TRACE ]] && set -x
   help | ps:help)
-    cat<<EOF
-    ps <app>, List processes running in app container(s)
+    help_contet='    ps <app>, List processes running in app container(s)
```
typo?

michaelshobbs commented on plugins/shell/commands:
```diff
@@ -4,9 +4,13 @@ set -eo pipefail; [[ $DOKKU_TRACE ]] && set -x
-    cat<<EOF
-    shell, Spawn dokku shell
-EOF
+    help_contnet='    shell, Spawn dokku shell'
```
typo?

michaelshobbs commented on Makefile:
```diff
@@ -57,10 +57,8 @@ copyfiles:
-		ln -s ${CORE_PLUGINS_PATH}/available/$$plugin ${PLUGINS_PATH}/available; \
 		find /var/lib/dokku/ -xtype l -delete;\
 		PLUGIN_PATH=${CORE_PLUGINS_PATH} plugn enable $$plugin ;\
-		PLUGIN_PATH=${PLUGINS_PATH} plugn enable $$plugin ;\
```
I seem to be missing where we enable plugins on install. Can you point me there? I **think** this is also executed in the the deb package.

michaelshobbs commented:
Can we make the split between community and core output without wrapping the plugn execution, and changing the plugin path variables?

michaelshobbs commented on plugins/git/commands:
```diff
@@ -54,11 +54,11 @@ git_hook_cmd() {
+      trigger-plugn trigger receive-app $APP $newrev
     else
-      if [[ $(find "$PLUGIN_PATH"/enabled/*/receive-branch 2>/dev/null | wc -l) != 0 ]]; then
+      if [[ $(find "$PLUGIN_CORE_PATH"/enabled/*/receive-branch 2>/dev/null | wc -l) != 0 || $(find "$PLUGIN_COMMUNITY_PATH"/enabled/*/receive-branch 2>/dev/null | wc -l) != 0 ]]; then
```
Probably could do this with one conditional like this:
```
if [[ $(find "$DOKKU_LIB_ROOT"/enabled/*/receive-branch 2>/dev/null | wc -l) != 0 ]]; then
```

michaelshobbs commented:
Example of how to only get core plugin help
```
    CORE_PLUGIN_COMMANDS=$(find -L "$PLUGIN_CORE_ENABLED_PATH" -name commands 2> /dev/null || true)
    for plugin_command in $CORE_PLUGIN_COMMANDS; do
      $plugin_command help
    done | sort | column -c2 -t -s,
```

u2mejc commented:
I totally understand and I even created code that was similar originally. The problem is that the rest of dokku expects to find core plugins in /var/lib/dokku/plugins/.

If this wasn't related to the help PR, how would you feel about segregating the core / community plugins and remove coding calling plugn directly?

michaelshobbs commented:
The PR pedantics aside, I don't see how the added complexity is beneficial.

michaelshobbs commented:
I'm not following what's wrong with your original approach and similarly my code example. Can you elaborate?

u2mejc commented:
Totally, and I agree, complexity for the sake of complexity doesn't make sense. 

Today, the core plugins (plugins that are part of this repo) are located in `/var/lib/dokku/core-plugins/available/`. Both the deb and the Makefile then symlink the core plugins to `/var/lib/dokku/core-plugins/enabled/` using `plugn enabled`, but not before the script symlinks the contents of `/var/lib/dokku/core-plugins/available` to `/var/lib/dokku/plugins/available` and then enabling them with `plugn enable` in both places. Take a look at this code block from debin/postinst for context:
```
    echo "Enabling all core plugins"
    find ${DOKKU_LIB_ROOT}/core-plugins/available -mindepth 1 -maxdepth 1 -type d -printf '%f\n' | while read -r plugin; do
      if [ ! -d ${DOKKU_LIB_ROOT}/plugins/available/$plugin ]; then
        ln -s ${DOKKU_LIB_ROOT}/core-plugins/available/$plugin ${DOKKU_LIB_ROOT}/plugins/available/$plugin;
        PLUGIN_PATH=${DOKKU_LIB_ROOT}/core-plugins plugn enable $plugin
        PLUGIN_PATH=${DOKKU_LIB_ROOT}/plugins plugn enable $plugin
      fi
    done
```

Related, in the Makefile, anytime `make copyfiles` runs, including install, all of the user's "community" plugins are spontaneously enabled. I've seen this has the potential to leave dokku in a broken state if a community plugin is broken.

The reason for this current state of file hierarchy may become apparent when looking at the code, parts of Dokku still expect all plugins (core or otherwise) to be found in /var/lib/dokku/plugins/ today. Both new world and old world are intertangled commonly like this:
```
source "$PLUGIN_CORE_AVAILABLE_PATH/common/functions"
source "$PLUGIN_AVAILABLE_PATH/checks/functions"
```
 Because of this assumption, it is difficult to make a clean division from what is core and what is community. Reading the symlink path like in the code example, would be one way to skin this cat. But it increases the glue code throughout the code base. We could just wrap up that glue in a common function to return if a passed plugin is core or not for each call, but we don't resolve for the human condition. Discerning sys admins may look at the directory system themselves, keeping that clean divide by not symlinking core plugins into the same directory as community plugins will help in avoiding the confusion of human error.

michaelshobbs commented:
Understood. I don't want to get too much in to the history of this decision but see the following for context
#1205 #1425 #1228 
To be clear, I'm also not debating that this PR is functional. 

`plugn` requires `PLUGIN_PATH` to be set in order to execute properly; as you know. In this PR we are now forced to run `plugn` twice; once for each set of plugins. How is this additional complexity justified? More to your point about increased "glue code throughout the code base", the symlink check only needs to be used when calling help. Admittedly we also do something similar in `execute_dokku_cmd()`. To that point, we could make this a common function in the future. (i.e. get_core_plugins or something like that.)  

I strongly think the symlink check instead of the reorganization (yet again) of the plugin directory structure is the most efficient and effective path to making this particular change.

josegonzalez commented:
I defer to @michaelshobbs on this pr.

josegonzalez commented on plugins/00_dokku-standard/subcommands/deploy:
```diff
@@ -1,8 +1,8 @@
-source "$PLUGIN_AVAILABLE_PATH/checks/functions"
-source "$PLUGIN_AVAILABLE_PATH/config/functions"
+source "$PLUGIN_CORE_AVAILABLE_PATH/checks/functions"
+source "$PLUGIN_CORE_AVAILABLE_PATH/config/functions"
```
Why these changes? We could previously override the implementations of core plugins, and now we can't.

josegonzalez commented on docs/docker-options.md:
````diff
@@ -5,8 +5,7 @@
 ```
-docker-options <app>                             Display app's Docker options for all phases
-docker-options <app> <phase(s)>                  Display app's Docker options for phase (comma-separated phase list)
+docker-options <app> [phase(s)], Display app'"'"'s Docker options for all phases (or comma separated phase list)
````
Mind adding the proper spacing instead?

u2mejc commented:
Refactored to drop the plugn wrapper and utilize most of @michaelshobbs example.

> Mind adding the proper spacing instead?

Done!

josegonzalez commented:
What about the change in sourcing for core plugins?

u2mejc commented:
@josegonzalez asked:
> What about the change in sourcing for core plugins?

After talking to Michael, we agreed that commit exceeded the scope of this "help" PR. I dropped that commit and just updated the help command to check it's symlink path.

If we're interested, we should probably open a separate RFC for plugin paths, managing plugins, and overriding / disabling core plugins.

josegonzalez commented on dokku:
```diff
@@ -157,17 +157,49 @@ execute_dokku_cmd() {
   *)
     execute_dokku_cmd "$@"
-    ;;
+  ;;
```
:(

josegonzalez commented on dokku:
```diff
@@ -157,17 +157,49 @@ execute_dokku_cmd() {
+        done | sort | column -c2 -t -s,
+      fi
+    fi
+  ;;
```
:( :(

josegonzalez commented:
Last two comments, and then I'm :+1: on this PR.

michaelshobbs commented:
Probably should update the plugin development docs to show how one might use the new help format

u2mejc commented:
will do, one sec

u2mejc commented:
Any other requests before we :shipit: ?

michaelshobbs commented on dokku:
```diff
@@ -157,13 +157,45 @@ execute_dokku_cmd() {
+    else
+      for core_plugin_command in $CORE_PLUGIN_COMMANDS; do
+        $core_plugin_command help
+      done | sort | sed -e '/^.*:/d' -e 's/\s[\[\<\-\(].*,/,/' | column -c2 -t -s,
```
Can you add a comment on what this `sed` regex is doing?

u2mejc replied:
Can do. 

Of note, I've intentionally been using less comments because the rest of the project has been using less comments. We use some declares to bubble up functions but very little comments anywhere else to explain code (especially would be nice for weird things like bash substitutions). If we're worried about bloat since the project is script based, we could have the comments stripped in the releases and only visible when checked out.

michaelshobbs replied:
Fair point around comments. My personal opinion on them is to only use them to describe functionality that isn't blatantly apparent. Is some cases we've wrapped the functionality in a "well-named" function. Dokku's consistency, aside I think it makes sense to use a comment here for now.

michaelshobbs commented:
Couple thoughts:
- I'd like to see some more descriptive documentation around the `help_content` variable and how it's used. Maybe mention in 0.5.x we expanded the usability of help and here's why you want to use it. Include a code block that shows how a user interfaces with `help` and what output looks like when you do and don't use `help_content`. 
- I'd prefer to avoid `\n` and instead just use multiple `echo` statements. I think this makes our core usage more "accessible" to community plugin devs as examples for implementation.

LMK if this doesn't make any sense

u2mejc commented:
_michaelshobbs commented 7 minutes ago_
>I'd like to see some more descriptive documentation around the help_content variable and how it's used. Maybe mention in 0.5.x we expanded the usability of help and here's why you want to use it. Include a code block that shows how a user interfaces with help and what output looks like when you do and don't use help_content.

Understood.

>I'd prefer to avoid \n and instead just use multiple echo statements. I think this makes our core usage more "accessible" to community plugin devs as examples for implementation.
LMK if this doesn't make any sense

Your reason is sound and I'll conform, but it doesn't mean I have to like it. :kissing_heart:

u2mejc commented:
Anything else after those? :smiley:

josegonzalez commented:
Nothing from me.

michaelshobbs commented on docs/docker-options.md:
````diff
@@ -5,8 +5,7 @@
 ```
-docker-options <app>                             Display app's Docker options for all phases
-docker-options <app> <phase(s)>                  Display app's Docker options for phase (comma-separated phase list)
+docker-options <app> [phase(s)]                  Display app's Docker options for all phases (or comma separated phase list)
````
+1 for POSIX standards :smile:

michaelshobbs commented:
Oh! My last bit would be to move the help commands to functions that include all local vars (or at least as much as possible); including `desc` and `cmd` vars like we do in other areas of the core plugin suite.

u2mejc commented:
@michaelshobbs When I need to edit multiple files, I use `vim $(grep blah blah)` or even `sed -i`. Your symlinks to files I've already edited in the repo are screwing with my flow. :cry:

michaelshobbs commented:
Feel free to use a "real" editor that understand symlinks or open an RFC issue to discuss. :smile:

josegonzalez commented:
Nano is pretty good.

michaelshobbs commented:
:trollface:

michaelshobbs commented on docs/development/plugin-triggers.md:
```diff
@@ -92,10 +92,10 @@ cd /root/dokku
 
-- Description: Used to aggregate all plugin `help` output. Your plugin should implement a `help` command in your `commands` file to take advantage of this plugin trigger. This must be implemented inside the `commands` plugin file.
-- Invoked by: `dokku help`
+- Description: Your plugin should implement a `help` command in your `commands` file to take advantage of this plugin trigger. `commands help` is used by `dokku help` to aggregate all plugins abbreviated `help` output. Implementing  `<PLUGIN_NAME>:help` in your `commands` file gives users looking for help, a more detail output. 'commands help' must be implemented inside the `commands` plugin file ...
```
typo change: "a more detailed output"
code block: `PLUGIN_NAME:help`
typo: "consistency"

u2mejc commented:
@michaelshobbs 
```diff
-detail-
+detailed+
-PLUGIN_NAME:help-
+`PLUGIN_NAME:help`+
-consitancty-
+consistency+
```

u2mejc commented:
:shipit: :shipit: :shipit:

josegonzalez commented:
:+1:

josegonzalez commented:
:clap:
